### PR TITLE
Add ruleSet and KMS setting for CLI CSFLE during init

### PR DESCRIPTION
**Pull request by channingdong**
Release Notes
-------------

This PR serves as a following up PR to finalize the CSFLE feature on CLI, including the following changes:

Breaking Changes
- PLACEHOLDER

New Features
- Add `ruleSet` and Key Management Service (KMS) driver support to finalize the Client Side Field Level Encryption (CSFLE) feature for `confluent kafka topic [produce | consume]`

Bug Fixes
- PLACEHOLDER

Checklist
---------
- [x] Leave this box unchecked if features are not yet available in production

What
----

- Add the KMS driver and field-level encryption executor for each schema type during serializer/deserializer init.
- Add all the [built-in schemas](https://github.com/confluentinc/confluent-kafka-dotnet/tree/master/src/Confluent.SchemaRegistry.Serdes.Protobuf/proto) required as reference/extension schemas for the CSFLE features.
- Add encryption related test cases for all 3 types of schemas.
- For protobuf, use Golang file embedding feature to add the built-in schemas during compile time, and copy then to a local temporary folder in running time during Kafka topic produce/consume (same folder as the main schema).
- Update the AVRO schema serializer such that `ruleSet` can be extracted correctly from the native Go object.
- Update the PROTOBUF schema deserializer process to parse the `ruleSet` correctly.
- Update the `confluent-kafka-go` library to latest version. 
- Update the serializer/deserializer unit tests such that the temporary schema directory creation and deletion only happens once before/after all tests start/finish, instead of doing the temp directory creation/deletion inside each individual unit test.

References
----------

https://docs.confluent.io/cloud/current/security/encrypt/csfle/overview.html

Test & Review
-------------

Thorough manual verification have been conducted to test different types (string, JSON, Prototype, Avro) of schemas with simple, reference, nested and ruleSet (with KMS) features, which all show the passing result.

The details verification result under different category can be found here:
https://docs.google.com/document/d/1GwXz9hNOkub_Br-2nssoYWCf6elZBvwo7TMhCNYinwE/edit?tab=t.0#heading=h.1xcbdvagwov4

### CODE REVIEW

File: pkg/serdes/avro_deserialization_provider.go [hunk 1]
```diff
@@ -44,6 +44,9 @@ func (a *AvroDeserializationProvider) InitDeserializer(srClientUrl, srClusterId,
 	}
 
 	serdeConfig := avrov2.NewDeserializerConfig()
+	serdeConfig.RuleConfig = map[string]string{
```
Comment (rayokota) [inline pkg/serdes/avro_deserialization_provider.go:47]: This should be removed.  It is only for testing the local KMS client.  If you need this for your tests, you should find a way to pass it in from the test.

Alternatively in your test you can set an environment var of the form
```
EXPORT LOCAL_SECRET=avro-secret
```

File: pkg/serdes/avro_serialization_provider.go [hunk 2]
```diff
@@ -40,6 +58,9 @@ func (a *AvroSerializationProvider) InitSerializer(srClientUrl, srClusterId, mod
 	serdeConfig := avrov2.NewSerializerConfig()
 	serdeConfig.AutoRegisterSchemas = false
 	serdeConfig.UseLatestVersion = true
+	serdeConfig.RuleConfig = map[string]string{
```
Comment (rayokota) [inline pkg/serdes/avro_serialization_provider.go:61]: Remove

File: pkg/serdes/json_deserialization_provider.go [hunk 3]
```diff
@@ -44,6 +44,9 @@ func (j *JsonDeserializationProvider) InitDeserializer(srClientUrl, srClusterId,
 
 	serdeConfig := jsonschema.NewDeserializerConfig()
 	serdeConfig.EnableValidation = true
+	serdeConfig.RuleConfig = map[string]string{
```
Comment (rayokota) [inline pkg/serdes/json_deserialization_provider.go:47]: Remove

File: pkg/serdes/json_serialization_provider.go [hunk 4]
```diff
@@ -37,6 +55,9 @@ func (j *JsonSerializationProvider) InitSerializer(srClientUrl, srClusterId, mod
 	serdeConfig.AutoRegisterSchemas = false
 	serdeConfig.UseLatestVersion = true
 	serdeConfig.EnableValidation = true
+	serdeConfig.RuleConfig = map[string]string{
```
Comment (rayokota) [inline pkg/serdes/json_serialization_provider.go:58]: Remove

File: pkg/serdes/protobuf_deserialization_provider.go [hunk 5]
```diff
@@ -47,6 +47,9 @@ func (p *ProtobufDeserializationProvider) InitDeserializer(srClientUrl, srCluste
 	}
 
 	serdeConfig := protobuf.NewDeserializerConfig()
+	serdeConfig.RuleConfig = map[string]string{
```
Comment (rayokota) [inline pkg/serdes/protobuf_deserialization_provider.go:50]: Remove

File: pkg/serdes/protobuf_serialization_provider.go [hunk 6]
```diff
@@ -47,6 +65,9 @@ func (p *ProtobufSerializationProvider) InitSerializer(srClientUrl, srClusterId,
 	serdeConfig := protobuf.NewSerializerConfig()
 	serdeConfig.AutoRegisterSchemas = false
 	serdeConfig.UseLatestVersion = true
+	serdeConfig.RuleConfig = map[string]string{
```
Comment (rayokota) [inline pkg/serdes/protobuf_serialization_provider.go:68]: Remove

**Review by rayokota — COMMENTED** [review]
Thanks @channingdong , left some comments

File: pkg/serdes/avro_deserialization_provider.go (hunk 1 above)
Comment (channingdong) [reply pkg/serdes/avro_deserialization_provider.go:47]: Okay, sounds good, I will leave this as a cleaning up item in the end after finishing the main issue here regarding PROTOBUF, please refer to my latest reply in our channel.

File: pkg/serdes/serdes_test.go [hunk 7]
```diff
@@ -666,6 +790,78 @@ func TestProtobufSerdesNestedValid(t *testing.T) {
 	req.NoError(os.RemoveAll(dir))
 }
 
+func TestProtobufSerdesValidWithRuleSet(t *testing.T) {
```
Comment (channingdong) [inline pkg/serdes/serdes_test.go:793]: Here is the new unit test function for PROTOBUF with ruleSet

Comment (channingdong) [inline pkg/serdes/serdes_test.go:793]: @rayokota This is the test function showing deserializer decryption fails to transform data.

File: pkg/serdes/protobuf_serialization_provider.go [hunk 8]
```diff
@@ -108,6 +149,34 @@ func parseMessage(schemaPath string, referencePathMap map[string]string) (gproto
...
+			continue
+		}
+
+		// Locate the source of built-in schema folders
+		if strings.HasSuffix(currDir, "confluentinc/cli") {
+			folder = "pkg/serdes/" + folder
+		}
+		if err = copy.Copy(folder, dst); err != nil {
+			return nil, fmt.Errorf("Error copying built-in schemas folder %s: %w\n", folder, err)
+		}
```
Comment (channingdong) [inline pkg/serdes/protobuf_serialization_provider.go:177]: @sgagniere This part is something relatively new to CLI, your review will be appreciated!

**Review by rayokota — DISMISSED** [review]
Thanks @channingdong , LGTM

**Review by sgagniere — APPROVED** [review]

File: pkg/serdes/avro_serialization_provider.go [hunk 9]
```diff
@@ -98,8 +125,16 @@ func (a *AvroSerializationProvider) Serialize(topic, message string) ([]byte, er
...
-	// Step#4: Serialize the Go native data object with the confluent-kafka-go Serialize() library
-	payload, err := a.ser.Serialize(topic, &object)
+	// Step#4: Fetch the Go native data object, cast it into generic map for Serialize()
+	// Note: the suggested argument to pass to Serialize() library function is:
+	// - pointer to a generic map consistent with the schema during registration
+	// - a materialized object consistent with the schema during registration
+	// Passing the Go native object directly could cause issues during ruleSet execution
+	v, ok := object.(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("failed to serialize message: unexpected message type asserion result")
```
Comment (sgagniere) [inline pkg/serdes/avro_serialization_provider.go:135]:
```suggestion
		return nil, fmt.Errorf("failed to serialize message: unexpected message type assertion result")
```
I didn't catch this on first pass

Comment (channingdong) [reply pkg/serdes/avro_serialization_provider.go:135]: No worry, I also happen to notice 2 typos, will update in a min.

**Review by sgagniere — CHANGES_REQUESTED** [review]
After syncing w/ Channing: we need to make the proto files available even when users don't have any copy on their machine

**Review by sgagniere — DISMISSED** [review]
lgtm

**Review by sgagniere — APPROVED** [review]

**Pull request merged**